### PR TITLE
Add httpapi VMware REST VmwareRestModule

### LINKED ISSUE #60913: VMware: Add httpapi VMware REST VmwareRestModule
##### SUMMARY
Implementing httpapi plugin-based REST modules for VMware.  Laying out initial directory structure and VmwareRestModule extended from AnsibleModule to support the connector and handle communications.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
- VmwareRestModule.py
- vmware_appliance_access_info.py
- vmware_appliance_health_info.py
- vmware_cis_category_info.py
- vmware_core_info.py

##### ADDITIONAL INFORMATION
The AnsibleModule class was extended to create the VmwareRestModule class that implements comms with the httpapi plugin. It also provides debugging output, url and filtering generation by object, dynamic handling of HTTP return codes that can be tailored to each module's needs, and support for multiple VMware REST APIs including, but not limited to appliance, cis, content, vapi, vcenter, and vrops.

Using the 'appliance' API:
```yaml
- hosts: all
  connection: httpapi
  gather_facts: false
  vars:
    ansible_network_os: vmware
    ansible_host: vcenter.my.domain
    ansible_user: administrator@vsphere.local
    ansible_httpapi_password: "SomePassword"
    ansbile_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: false
  tasks:

  - name: Get all access modes information
    vmware_appliance_access_info:

  - name: Get ssh access mode information
    vmware_appliance_access_info:
      access_mode: ssh
```
Using the 'vcenter' API:
```yaml
- hosts: all
  connection: httpapi
  gather_facts: false
  vars:
    ansible_network_os: vmware
    ansible_host: vcenter.my.domain
    ansible_user: some.user
    ansible_httpapi_password: "SomePassword"
    ansible_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: false
  tasks:

  - name: get vm information
    vmware_core_info:
      object_type: vm
      filters:
        - names: "some_vm"
```

### PR DESCRIPTION
**Pull request by n3pjk**
##### SUMMARY
Implementing httpapi plugin-based REST modules for VMware.  Laying out initial directory structure and VmwareRestModule extended from AnsibleModule to support the connector and handle communications.  These modules must all reside in modules/cloud/vmware-httpapi, as directories can only go two deep under modules, and vmware is already under modules/cloud.

Fixes
- #60913

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- VmwareRestModule.py
- vmware_appliance_access_info.py
- vmware_appliance_health_info.py
- vmware_cis_category_info.py
- vmware_core_info.py

##### ADDITIONAL INFORMATION
The AnsibleModule class was extended to create the VmwareRestModule class that implements comms with the httpapi plugin. It also provides debugging output, url and filtering generation by object, dynamic handling of HTTP return codes that can be tailored to each module's needs, and support for multiple VMware REST APIs including, but not limited to appliance, cis, content, vapi, vcenter, and vrops.

Using the 'appliance' API:
```yaml
- hosts: all
  connection: httpapi
  gather_facts: false
  vars:
    ansible_network_os: vmware
    ansible_host: vcenter.my.domain
    ansible_user: administrator@vsphere.local
    ansible_httpapi_password: "SomePassword"
    ansbile_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: false
  tasks:

  - name: Get all access modes information
    vmware_appliance_access_info:

  - name: Get ssh access mode information
    vmware_appliance_access_info:
      access_mode: ssh
```
Using the 'vcenter' API:
```yaml
- hosts: all
  connection: httpapi
  gather_facts: false
  vars:
    ansible_network_os: vmware
    ansible_host: vcenter.my.domain
    ansible_user: some.user
    ansible_httpapi_password: "SomePassword"
    ansible_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: false
  tasks:

  - name: get vm information
    vmware_core_info:
      object_type: vm
      filters:
        - names: "some_vm"
```

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware-httpapi/vmware_core_info.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    - Valid filters for resource_pool object type - resource_pools, parent_resource_pools, names, datacenters, hosts, clusters.
+    - Valid filters for virtual_machine object type - folders, resource_pools, power_states, vms, names, datacenters, hosts, clusters.
+    - content_library, local_library, subscribed_library, content_type, tag, category does not take any filters.
+    default: []
+    type: list
+'''
+
+EXAMPLES = r'''
+- name: Get Datacenter object information
+  vmware_get_info:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware-httpapi/vmware_core_info.py]:
```suggestion
  vmware_core_info:
```

File: lib/ansible/modules/cloud/vmware-httpapi/vmware_core_info.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    object_type: datacenter
+    filters:
+      - names: Asia-Datacenter1
+  register: datacenter_result
+
+- set_fact:
+    datacenter_obj: "{{ datacenter_result.object_info.value[0]['datacenter'] }}"
+
+- name: Get all clusters from Asia-Datacenter1
+  vmware_get_info:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware-httpapi/vmware_core_info.py]:
```suggestion
  vmware_core_info:
```

**Review by Akasurde — COMMENTED** [review]
I am testing this module on my side. Will post more comments once I am done.

**Comment by n3pjk** [comment]
I had not done any tests for httpapi yet.  From the test results, any modules in cloud/vmware-httpapi need to use the plugin and not pyvmomi.

**Comment by Akasurde** [comment]
@n3pjk I added few changes from my first round of testing. Please merge https://github.com/n3pjk/ansible/pull/3 in this.

**Comment by Akasurde** [comment]
@n3pjk I added few changes from my first round of testing. Please merge https://github.com/n3pjk/ansible/pull/3 in this.

**Comment by n3pjk** [comment]
I'm close to eliminating all errors in CI. Waiting now to see if my latest changes green light the PR. Regarding a test play, look at the example documented in vmware_appliance_access_info.py. The top of the play, down to "tasks:" is the same for all appliance api plays. For all other apis, use your vcenter user creds for ansible_user and ansible_httpapi_password.

When you have time we need to go over how I can create tests for httpapi-based modules. I figured out why the CI kept wanting to use pyvmomi with httpapi. I had created module_util/vmware and there was also a module_util/vmware.py and that caused the tests to try to import Pyvmomi et al from the directory, not the file! Silly snake. Lol
________________________________

**Comment by n3pjk** [comment]
I do use an inventory file, but it only sets the Service Now region. The command line I use is "ansible-playbook - i httpapi_appliance.yml - vvvvv" Where the yaml file is as specified in the example mentioned previously. Make sure you have the right password for administrator@vsphere.local, assuming that account is accessible to you. You must use that account for the appliance API. The vCenter API uses an arbitrary cred that has admin privileges to vCenter. Same for the cis API. I'm available to Zoom tonight or tomorrow if you want.

Maybe then you can show me the right way to set up tests for these beasties. Also, should I just use molecule to run these tests locally?
________________________________

**Comment by n3pjk** [comment]
I will address these two issues shortly.
________________________________

**Comment by n3pjk** [comment]
I have been working on trying to understand the OS-specific failures and instabilities, but it looks like they are not related to my code.  They also appear to be inconsistent.  Is this acceptable to be merged?

**Comment by n3pjk** [comment]
Is there anything else I can do?
________________________________

**Comment by n3pjk** [comment]
I just reran CI and, miraculously, all tests passed!

**Comment by n3pjk** [comment]
Try "hosts: all" instead of "hosts: local host". That's how my plays are configured.
________________________________

**Comment by goneri** [comment]
> Try "hosts: all" instead of "hosts: local host". That's how my plays are configured.

My inventory is empty, so `all` just do nothing in this case.

**Comment by goneri** [comment]
I tried to call `vmware_core_info` without any filter, and so I dropped the `filters` key. This raised an error:

```
The full traceback is:
Traceback (most recent call last):
  File "/home/goneri/.ansible/tmp/ansible-local-30173yhhxjxtw/ansible-tmp-1568211859.585594-269761304020390/AnsiballZ_vmware_core_info.py", line 102, in <module>
    _ansiballz_main()
  File "/home/goneri/.ansible/tmp/ansible-local-30173yhhxjxtw/ansible-tmp-1568211859.585594-269761304020390/AnsiballZ_vmware_core_info.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/goneri/.ansible/tmp/ansible-local-30173yhhxjxtw/ansible-tmp-1568211859.585594-269761304020390/AnsiballZ_vmware_core_info.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.cloud.vmware_httpapi.vmware_core_info', init_globals=None, run_name='__main__', alter_sys=False)
  File "/usr/lib64/python2.7/runpy.py", line 192, in run_module
    fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/tmp/ansible_vmware_core_info_payload_kfGiiy/ansible_vmware_core_info_payload.zip/ansible/modules/cloud/vmware_httpapi/vmware_core_info.py", line 117, in <module>
  File "/tmp/ansible_vmware_core_info_payload_kfGiiy/ansible_vmware_core_info_payload.zip/ansible/modules/cloud/vmware_httpapi/vmware_core_info.py", line 110, in main
  File "/tmp/ansible_vmware_core_info_payload_kfGiiy/ansible_vmware_core_info_payload.zip/ansible/module_utils/vmware_httpapi/VmwareRestModule.py", line 643, in get_url_with_filter
  File "/tmp/ansible_vmware_core_info_payload_kfGiiy/ansible_vmware_core_info_payload.zip/ansible/module_utils/vmware_httpapi/VmwareRestModule.py", line 636, in get_url
TypeError: cannot concatenate 'str' and 'NoneType' objects
```

My playbook:

```
  - name: get vm information
    vmware_core_info:
      object_type: host
```

**Comment by n3pjk** [comment]
I'll clean that up shortly.
________________________________

**Review by goneri — APPROVED** [review]

**Pull request merged**